src/som/interpreter/nodes/literals/ObjectLiteralNode.java

```diff
@@ -0,0 +1,77 @@
+package som.interpreter.nodes.literals;
+
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.source.SourceSection;
+
+import som.compiler.MixinDefinition;
+import som.interpreter.InliningVisitor;
+import som.interpreter.nodes.ExpressionNode;
+import som.interpreter.nodes.MessageSendNode.AbstractMessageSendNode;
+import som.vmobjects.SClass;
+import som.vmobjects.SObjectWithClass;
+
+
+public class ObjectLiteralNode extends LiteralNode {
+
+  private MixinDefinition mixinDef;
+  private RootCallTarget superClassResolver;
+  private SObjectWithClass outer;
+  private Object superclassAndMixins;
+
+  @Child private ExpressionNode outerRead;
+  @Child private AbstractMessageSendNode newMessage;
+
+  /**
+   * An object literal node. When invoked this node will instantiate a new object literal class
+   * (defined by this node's `MixinDefinition`) that has the current activation as its enclosing
+   * scope. Once this class has been created, this node will instantiate and return a new object
+   * from that class.
+   */
+  public ObjectLiteralNode(final MixinDefinition mixinDefiniton, final ExpressionNode outerRead,
+      final ExpressionNode newMessage, final SourceSection source) {
+    super(source);
+    this.mixinDef = mixinDefiniton;
+    this.superClassResolver = mixinDefiniton.getSuperclassAndMixinResolutionInvokable().createCallTarget();
+    this.outer = null;
+    this.superclassAndMixins = null;
+
+    this.outerRead = outerRead;
+    this.newMessage = (AbstractMessageSendNode) newMessage;
+  }
+
+  /**
+   * This method caches the outer object and superclass and mixins used to instantiate the object
+   * literal. Note that this method is only called once when this node is invoked for the first
+   * time.
+   */
+  private void cacheOuterAndSuperClass(final VirtualFrame frame) {
+    assert outer == null && superclassAndMixins == null :
+      "The outer object and superclass and mixins can only be cached once.";
+
+    outer = (SObjectWithClass) outerRead.executeGeneric(frame);
+    superclassAndMixins = superClassResolver.call(outer);
```

Comment:
And this is also not safe.

Please see `ClassInstantiationNode` again, there are guards to cache things.
Perhaps you can adapt `ClassInstantiationNode` to support your case so you don't need to roll your own?